### PR TITLE
Iter9

### PR DESCRIPTION
**Pull request by AnatolySnegovskiy**

### CODE REVIEW

File: internal/services/agent/metrics.go [hunk 1]
```diff
@@ -1,31 +1,55 @@
 package agent
 
 import (
+	"bytes"
+	"compress/gzip"
 	"context"
 	"fmt"
+	"github.com/AnatolySnegovskiy/metric/internal/services/dto"
```
Comment (arttet) [inline internal/services/agent/metrics.go:8]: CHANGE REQUEST: please sort these imports

File: internal/services/agent/metrics.go [hunk 2]
```diff
@@ -1,31 +1,55 @@
...
-				return err
+			if storageType == "counter" {
+				iv := int64(metric)
+				metricDto.Delta = &iv
+			} else {
+				metricDto.Value = &metric
 			}
 
-			defer resp.Body.Close()
+			var buf bytes.Buffer
```
Comment (arttet) [inline internal/services/agent/metrics.go:31]: SUGGESTION: it should be in the middleware

File: internal/services/agent/metrics.go [hunk 3]
```diff
@@ -1,31 +1,55 @@
...
+			_ = gw.Close()
 
-			return nil
+			url := fmt.Sprintf("http://%s/update/", a.sendAddr)
+			req, _ := http.NewRequestWithContext(ctx, http.MethodPost, url, &buf)
+			req.Header.Set("Content-Encoding", "gzip")
+			req.Header.Set("Content-Type", "application/json")
+			resp, _ := a.client.Do(req)
+
+			if resp != nil {
```
Comment (arttet) [inline internal/services/agent/metrics.go:45]: CHANGE REQUEST: this is not Go style approach.

File: internal/services/server/handlers.go [hunk 4]
```diff
@@ -75,6 +117,57 @@ func (s *Server) showMetricNameHandlers(rw http.ResponseWriter, req *http.Reques
...
+	}
+
+	if metricDTO.MType == "gauge" {
+		metricDTO.Value = &metric
+	} else {
+		val := int64(metric)
+		metricDTO.Delta = &val
+	}
+
+	json, _ := easyjson.Marshal(metricDTO)
```
Comment (arttet) [inline internal/services/server/handlers.go:155]: CHANGE REQUEST: please check the error

File: internal/services/server/middleware.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+			r.RequestURI,
+			duration,
+			r.ContentLength,
+		)
+
+		s.logger.Infof("response status: %d; size: %d;", responseData.status, responseData.size)
+	})
+}
+
+func (s *Server) gzipResponseMiddleware(next http.Handler) http.Handler {
```
Comment (arttet) [inline internal/services/server/middleware.go:61]:
```suggestion
func (s *Server) gzipCompressMiddleware(next http.Handler) http.Handler {
```

File: internal/services/server/middleware.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+		defer gz.Close()
+		w.Header().Set("Content-Encoding", "gzip")
+		w.Header().Del("Content-Length")
+		w = &gzipResponseWriter{ResponseWriter: w, Writer: gz}
+
+		next.ServeHTTP(w, r)
+	})
+}
+
+func (s *Server) gzipRequestMiddleware(next http.Handler) http.Handler {
```
Comment (arttet) [inline internal/services/server/middleware.go:78]:
```suggestion
func (s *Server) gzipDecompressMiddleware(next http.Handler) http.Handler {
```

File: internal/services/server/server.go [hunk 7]
```diff
@@ -37,3 +49,65 @@ func (s *Server) setupRoutes() {
...
+			}
+		}
+	}
+
+	s.logger.Info("Metrics loaded: " + filePath)
+}
+
+func (s *Server) SaveMetricsToFile(filePath string) {
+	projectDir, _ := os.Getwd()
+	absoluteFilePath := filepath.Join(projectDir, filePath)
```
Comment (arttet) [inline internal/services/server/server.go:89]: CHANGE REQUEST: there is a bug here.

File: internal/services/server/server.go [hunk 8]
```diff
@@ -37,3 +49,65 @@ func (s *Server) setupRoutes() {
...
+
+	file, _ := os.Create(absoluteFilePath)
+	defer file.Close()
+	jsonData, _ := json.Marshal(s.storage.GetList())
+
+	_, _ = file.Write(jsonData)
+	s.logger.Info("Metrics saved: " + absoluteFilePath)
+}
+
+func loadMetricsFromFile(filePath string) map[string]map[string]map[string]float64 {
```
Comment (arttet) [inline internal/services/server/server.go:102]: P.S. https://refactoring.guru/design-patterns/memento/go/example

**Review by arttet — APPROVED** [review]

File: internal/services/server/middleware.go (hunk 6 above)
Comment (AnatolySnegovskiy) [reply internal/services/server/middleware.go:78]: ++

File: internal/services/server/middleware.go (hunk 5 above)
Comment (AnatolySnegovskiy) [reply internal/services/server/middleware.go:61]: ++

File: internal/services/agent/metrics.go (hunk 1 above)
Comment (AnatolySnegovskiy) [reply internal/services/agent/metrics.go:8]: сортировку делал goland ни че там не подсвечивает.

**Pull request merged**